[RPC] refactor autocombine into specific commands

Fuzzbawls opened the pull request:
The current `autocombine` command is limited to whole numbers only, and infringes on our RPC standards by not providing deterministic output even when successfully changing the setting.

This PR addresses the deterministic issue as well as makes the setting information available to `getwalletinfo`

JSKitty commented on doc/release-notes.md:
```diff
@@ -44,6 +44,44 @@ Notable Changes
 
+#### Show wallet's auto-combine settings in getwalletinfo
+
+`getwalletinfo` now has two additional return fields. `autocombine_enabled` (boolean) and `autocombine_threshold` (numeric) that will show the auto-combine threshold and weather or not it is currently enabled.
```
tiny nit: weather --> whether

Fuzzbawls replied:
fixed

random-zebra commented on src/wallet/walletdb.cpp:
```diff
@@ -637,6 +637,9 @@ bool ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue, CW
             ssValue >> pSettings;
             pwallet->fCombineDust = pSettings.first;
             pwallet->nAutoCombineThreshold = pSettings.second;
+            // originally saved as integer
+            if (pwallet->nAutoCombineThreshold < COIN)
+                pwallet->nAutoCombineThreshold *= COIN;
```
this will create issues if someone sets a threshold below 1 coin.
Maybe we should just use a different key in the database, e.g. `autocombinesettingsV2`.

Fuzzbawls replied:
It's currently (`master` branch) not possible to set the threshold below `1` as the `autocombine` RPC command only accepts whole numbers. So this conversion is "safe" in the sense of upgrading from an older wallet that had an int value set previously.

I think it would be acceptable to make it a local policy in the new command to also not allow values less than `1` as such a low threshold is basically "dust" and transaction fees would make it pretty uneconomic/unreasonable. Even further, the biggest use case for autocombine is to combine multiple 3 PIV masternode rewards into a larger single UTXO, so setting a threshold less than `3` wouldn't be effective in achieving this.

random-zebra replied:
Yes. My only concern was about *after* this change, when it could be possible to set a threshold of, say, `0.99`.
I'd be ok with introducing a `1 PIV` limit in the RPC.

random-zebra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3934,6 +3935,84 @@ UniValue autocombinerewards(const JSONRPCRequest& request)
+
+    RPCTypeCheck(request.params, {UniValue::VBOOL, UniValue::VNUM});
+
+    bool fEnable = false;
+    if (!request.params.empty())
+        fEnable = request.params[0].get_bool();
```
Params cannot be empty here, otherwise the help runtime_error would be thrown.

random-zebra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3934,6 +3935,84 @@ UniValue autocombinerewards(const JSONRPCRequest& request)
+        fEnable = request.params[0].get_bool();
+
+    CWalletDB walletdb(pwalletMain->GetDBHandle());
+    CAmount nThreshold = 0;
+
+    if (fEnable && request.params.size() > 1)
+        nThreshold = AmountFromValue(request.params[1]);
```
styling nit: can inline as
```cpp
const CAmount nThreshold = (fEnable && request.params.size() > 1) ? AmountFromValue(request.params[1]) : 0;
```

random-zebra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3934,6 +3935,84 @@ UniValue autocombinerewards(const JSONRPCRequest& request)
+
+UniValue getautocombinethreshold(const JSONRPCRequest& request)
+{
+    if (request.fHelp || request.params.size() != 0)
```
styling nit:
```suggestion
    if (request.fHelp || !request.params.empty())
```

random-zebra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3934,6 +3935,84 @@ UniValue autocombinerewards(const JSONRPCRequest& request)
+            HelpExampleCli("getautocombinethreshold", "") + HelpExampleRpc("getautocombinethreshold", ""));
+
+    UniValue result(UniValue::VOBJ);
+    result.pushKV("enabled", pwalletMain->fCombineDust);
+    result.pushKV("threshold", ValueFromAmount(pwalletMain->nAutoCombineThreshold));
```
we should lock `cs_wallet` before reading `fCombineDust` and `nAutoCombineThreshold`

random-zebra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3934,6 +3935,84 @@ UniValue autocombinerewards(const JSONRPCRequest& request)
+
+            "\nExamples:\n" +
+            HelpExampleCli("getautocombinethreshold", "") + HelpExampleRpc("getautocombinethreshold", ""));
+
```
Since we access pwalletMain, it would be good to `EnsureWallet()` here

random-zebra reviewed (commented):
Concept ACK. Few points.

Fuzzbawls commented:
Rebased and updated to address feedback.

The setter now doesn't allow a threshold value below `1`, and also checks to ensure that a threshold value is given if `fEnable` is `True`.

I've also added a simple functional test for the two new commands

random-zebra commented on src/wallet/rpcwallet.cpp:
```diff
@@ -3934,6 +3938,92 @@ UniValue autocombinerewards(const JSONRPCRequest& request)
+    CWalletDB walletdb(pwalletMain->GetDBHandle());
+    CAmount nThreshold = 0;
+
+    if (fEnable && request.params.size() > 1) {
```
nit: no need to check the params size again, if we didn't throw at line 3971 we are good. 
We could refactor this a bit, and do both checks here, e.g.
```cpp
if (fEnable) {
    if (request.params.size() < 2) {
        throw JSONRPCError(RPC_INVALID_PARAMETER, "Missing threshold value");
    }
    nThreshold = AmountFromValue(request.params[1]);
    ...
}
```

Fuzzbawls replied:
needed to rebase this anyways, so took care of this while i was at it

random-zebra reviewed (commented):
Code review ACK with a nit.

random-zebra reviewed (approved):
utACK 84f9cab64c062f96e07baa3528836f11502afb50

random-zebra reviewed (approved):
re-utACK 2444de9bc19ed95872853b52f2d7e865fb7b2d29 after rebase

furszy reviewed (approved):
utACK 2444de9bc19ed95872853b52f2d7e865fb7b2d29 and merging..
